lib/query/token-visitors.js

```diff
@@ -0,0 +1,188 @@
+        if (isFunction(replacement)) {
+            let token = cloneDeep(this.tokenListVisitor.tokens[this.index])
+            replacement = replacement(token)
+        }
+        if (isString(replacement)) replacement = this.tokenListVisitor.tokenize(replacement);
+        return this.tokenListVisitor._replaceAtIndex(this.index, replacement)
+    }
+    insertBefore(tokens) {
+        if (isString(tokens)) tokens = this.tokenListVisitor.tokenize(tokens);
+        if (isEmpty(tokens)) throw new Error("Expected to find atleast one token");
+        return this.tokenListVisitor._insertAtIndex(this.index, tokens);
+    }
+    insertAfter(tokens) {
+        if (isString(tokens)) tokens = this.tokenListVisitor.tokenize(tokens);
+        if (isEmpty(tokens)) throw new Error("Expected to find atleast one token");
+        return this.tokenListVisitor._insertAtIndex(this.index + 1, tokens);
+    }
+    _normalizeSelectionPredicate(predicate, targetIndex) {
+        if (isNil(predicate)) return takeOneAt(targetIndex);
+        if (isString(predicate)) return (t) => t === predicate;
+        if (isFunction(predicate)) return predicate;
+        throw new Error(`Invalid predicate supplied: ${predicate}. Expected function or token`);
+    }
+    select(until) {
+        const predicate = this._normalizeSelectionPredicate(until, this.index);
+        return this.tokenListVisitor._selectRange(this.index, 1, predicate);
+    }
+    selectPrev(until) {
+        const start = this.index - 1;
+        const predicate = this._normalizeSelectionPredicate(until, start)
+        return this.tokenListVisitor._selectRange(start, -1, predicate);
+    }
+    selectNext(until) {
+        const start = this.index + 1;
+        const predicate = this._normalizeSelectionPredicate(until, start)
+        return this.tokenListVisitor._selectRange(start, 1, predicate);
+    }
+}
+
+class TokenRangeVisitor {
+    constructor(startIndex, endIndex, tokenListVisitor) {
+        this.startIndex = startIndex;
+        this.endIndex = endIndex;
+        this.tokenListVisitor = tokenListVisitor;
+    }
+    first() {
+        return new TokenItemVisitor(this.startIndex, this.tokenListVisitor);
+    }
+    last() {
+        return new TokenItemVisitor(this.endIndex, this.tokenListVisitor);
+    }
+    extendLeft(until) {
+        const rangeVisitor = this.tokenListVisitor._selectRange(this.startIndex, -1, until);
+        if (!isNil(rangeVisitor)) {
+            rangeVisitor.endIndex = this.endIndex;
+        }
+        return rangeVisitor;
+    }
+    extendRight(until) {
+        const rangeVisitor = this.tokenListVisitor._selectRange(this.endIndex, 1, until);
+        if (!isNil(rangeVisitor)) {
+            rangeVisitor.startIndex = this.startIndex;
+        }
+        return rangeVisitor
+    }
+    replace(replacement) {
+        if (isFunction(replacement)) {
+            let tokens = cloneDeep(this.tokenListVisitor.tokens.slice(this.startIndex, this.endIndex))
```

Comment:
why use `let`? You intend to modify this variable at some point?

Reply:
Updated to const.